<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Откройте-и-изучите-файл" data-toc-modified-id="Откройте-и-изучите-файл-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откройте и изучите файл</a></span><ul class="toc-item"><li><span><a href="#Импортируем-библиотеки" data-toc-modified-id="Импортируем-библиотеки-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импортируем библиотеки</a></span></li><li><span><a href="#Открыть-и-изучить-файл" data-toc-modified-id="Открыть-и-изучить-файл-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Открыть и изучить файл</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Разобьем-данные-на-выборки" data-toc-modified-id="Разобьем-данные-на-выборки-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разобьем данные на выборки</a></span><ul class="toc-item"><li><span><a href="#Обучающая" data-toc-modified-id="Обучающая-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обучающая</a></span></li><li><span><a href="#Валидационная-и-тестовая" data-toc-modified-id="Валидационная-и-тестовая-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Валидационная и тестовая</a></span></li></ul></li><li><span><a href="#Исследуйте-модели" data-toc-modified-id="Исследуйте-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследуйте модели</a></span><ul class="toc-item"><li><span><a href="#Logistic-Regression-(Логическая-регрессия)" data-toc-modified-id="Logistic-Regression-(Логическая-регрессия)-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Logistic Regression (Логическая регрессия)</a></span></li><li><span><a href="#Decision-Tree-(Дерево-решений)" data-toc-modified-id="Decision-Tree-(Дерево-решений)-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Decision Tree (Дерево решений)</a></span></li><li><span><a href="#Random-Forest-(Случайный-лес)" data-toc-modified-id="Random-Forest-(Случайный-лес)-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Random Forest (Случайный лес)</a></span></li></ul></li><li><span><a href="#Проверьте-модель-на-тестовой-выборке" data-toc-modified-id="Проверьте-модель-на-тестовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверьте модель на тестовой выборке</a></span></li><li><span><a href="#(бонус)-Проверьте-модели-на-адекватность" data-toc-modified-id="(бонус)-Проверьте-модели-на-адекватность-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>(бонус) Проверьте модели на адекватность</a></span></li><li><span><a href="#Вывод-по-проекту" data-toc-modified-id="Вывод-по-проекту-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод по проекту</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Описание проекта

Оператор мобильной связи «Мегалайн» выяснил: многие клиенты пользуются архивными тарифами. Они хотят построить систему, способную проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра».


В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.


Постройте модель с максимально большим значением accuracy. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверить accuracy на тестовой выборке.

# План по выполнению проекта
- Открыть файл с данными и изучить его. Путь к файлу: /datasets/users_behavior.csv. Скачать датасет
- Разделить исходные данные на обучающую, валидационную и тестовую выборки.
- Исследовать качество разных моделей, меняя гиперпараметры. Кратко написать выводы исследования.
- Проверьте качество модели на тестовой выборке.
- Дополнительное задание: проверить модели на вменяемость. 

# Описание данных
Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц. Известно:
- сalls — количество звонков,
- minutes — суммарная длительность звонков в минутах,
- messages — количество sms-сообщений,
- mb_used — израсходованный интернет-трафик в Мб,
- is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы. Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится.

Необходимо Построить модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверить *accuracy* на тестовой выборке самостоятельно.

## Откройте и изучите файл

### Импортируем библиотеки

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import roc_curve
from sklearn.metrics import auc
from tqdm import tqdm
import matplotlib.pyplot as plt

### Открыть и изучить файл

In [2]:
df = pd.read_csv('/datasets/users_behavior.csv')

In [3]:
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


Можно поменять тип поля messages и calls на целое число (int), т.к. дробных значений сообщений и звонков не бывает.

In [5]:
df['messages']=df['messages'].astype("int")

In [6]:
df['calls']=df['calls'].astype("int")

Проверим дубликаты

In [7]:
df.duplicated().sum()

0

### Вывод

Таблица содержит данные о 3214 пользователях которые перешли на данные тарифы, пропусков и дубликатов не обнаружено. Заменён тип данных в столбце 'messages' и 'calls' на целочисленый.

## Разобьем данные на выборки

In [8]:
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']

Спрятанной тестовой выборки нет. Поэтому, данные разобьем на три части: обучающую, валидационную и тестовую. Исходные данные разбивают в соотношении 3:1:1.

### Обучающая

In [9]:
features_train, features_test, target_train, target_test = train_test_split(features, 
                                                                            target, 
                                                                            test_size=0.4, 
                                                                            random_state=12345
                                                                           ) 

Размер выборки обучающей выборки

In [10]:
print(features_train.shape)
print(target_train.shape)

(1928, 4)
(1928,)


### Валидационная и тестовая

In [11]:
features_valid, features_test, target_valid, target_test = train_test_split(features_test, 
                                                                            target_test, 
                                                                            test_size=0.5, 
                                                                            random_state=12345
                                                                           ) 

Размер валидационной выборки

In [12]:
print(features_valid.shape)
print(target_valid.shape)

(643, 4)
(643,)


Размер тестовой выборки

In [13]:
print(features_test.shape)
print(target_test.shape)

(643, 4)
(643,)


## Исследуйте модели

Возьмем 3 модели:

- Logistic Regression (Логическая регрессия)
- Decision Tree (Дерево решений)
- Random Forest (Случайный лес)


### Logistic Regression (Логическая регрессия)

In [14]:
model_lr = LogisticRegression(random_state=12345).fit(features_train, target_train)
predictions_valid_lr = model_lr.predict(features_valid)
accuracy_lr = accuracy_score(predictions_valid_lr, target_valid)
print('Accuracy LogisticRegression:', accuracy_lr, end ='')

Accuracy LogisticRegression: 0.7107309486780715

Вывод
Accuracy модели LogisticRegression равно 0.7107309486780715

### Decision Tree (Дерево решений)

In [15]:
best_model_dtc = None
best_depth_dtc = 0
best_result_dtc = 0
for depth in tqdm(range(1, 11)):
    model_dtc = DecisionTreeClassifier(random_state=12345, max_depth = depth).fit(features_train, target_train)
    predictions_valid_dtc = model_dtc.predict(features_valid)
    result_dtc = accuracy_score(predictions_valid_dtc, target_valid)
    if best_result_dtc < result_dtc:
        best_model_dtc = model_dtc
        best_depth_dtc = depth
        best_result_dtc = result_dtc

100%|██████████| 10/10 [00:00<00:00, 136.10it/s]


In [16]:
print('best max_depth =', 
      best_depth_dtc, 
      '; ', 
      'best Accuracy = ', 
      best_result_dtc, 
      end =''
     ) 

best max_depth = 3 ;  best Accuracy =  0.7853810264385692

Вывод
Accuracy лучшей модели из десяти равно 0.7853810264385692, с гиперпараметром max_depth равным - 3, ресурсозатратность же равна, при заданных 10 иттераций, равна 150.06ит/сек.

### Random Forest (Случайный лес)

In [17]:
best_model_rfc = None
best_est_rfc = 0
best_depth_rfc = 0
best_result_rfc = 0
for est in tqdm(range(1, 11)):
    for depth in range(1, 11):
        model_rfc = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth = depth).fit(features_train, 
                                                                                                        target_train)
        predictions_valid_rfc = model_rfc.predict(features_valid)
        result_rfc = accuracy_score(predictions_valid_rfc, target_valid)
        if best_result_rfc < result_rfc:
            best_model_rfc = model_rfc
            best_est_rfc = est
            best_depth_rfc = depth
            best_result_rfc = result_rfc

100%|██████████| 10/10 [00:01<00:00,  6.29it/s]


In [18]:
print('best n_estimators =', 
      best_est_rfc, 
      '; ', 
      'best max_depth =', 
      best_depth_rfc, 
      '; ', 
      'best Accuracy = ', 
      best_result_rfc, 
      end =''
     ) 

best n_estimators = 8 ;  best max_depth = 8 ;  best Accuracy =  0.80248833592535

Вывод
Accuracy лучшей модели из десяти равно 0.80248833592535, с гиперпараметром n_estimators равным - 8 и max_depth равным 8, ресурсозатратность же равна, при заданных 10 иттераций, равна 5.59ит/сек.

Вывод
- наилучший результат Accuracy у модели решающего дерева - DecisionTree - Accuracy равен 0.7853810264385692, ресурсозатратность же при этом меньшая - при заданных 10 иттераций, равна 150.06ит/сек;
- наилучший результат Accuracy у модели случайного леса - RandomForest - 0.80248833592535, ресурсозатратность же, при заданных 10 иттераций, равна 5.59ит/сек.
- результат Accuracy у модели логистической регрессии - LogisticRegression - 0.7107309486780715.

## Проверьте модель на тестовой выборке

Наиболее эффективно себя показала модель случайного леса - RandomForest, поэтому проверим ее на тестовой выборке

In [19]:
model_rfc = RandomForestClassifier(random_state=12345, n_estimators = 8, max_depth = 8).fit(features_train, target_train)
predictions_test_rfc = model_rfc.predict(features_test)
result_test_rfc = accuracy_score(predictions_test_rfc, target_test)
print('Accuracy RandomForestClassifier:', result_test_rfc)

Accuracy RandomForestClassifier: 0.7962674961119751


Вывод
Accuracy у модели случайного леса - RandomForest на тестовой выборке равна 0.7962674961119751
На тестовой выборке выбранная модель показывает точность близкую к валидационной, но не больше. Это означает отсутствие проблем с переобучение модели.

## (бонус) Проверьте модели на адекватность

Для проверки используем модель DummyClassifier.

In [20]:
model_dc = DummyClassifier(strategy='most_frequent', random_state=12345)
model_dc.fit(features_train, target_train)
result_dc = model_dc.score(features_valid, target_valid)
print('Accuracy DummyClassifier:', result_dc)

Accuracy DummyClassifier: 0.7060653188180405


In [21]:
print('Accuracy RandomForest:', result_test_rfc)

Accuracy RandomForest: 0.7962674961119751


Вывод
Accuracy модели DummyClassifier ниже, чем у модели RandomForest - модель адекватна и эффективна.

## Вывод по проекту

- В проекте были изучены данные о поведении клиентов, которые уже перешли на эти тарифы.
- В данных не было обнаружено ошибок, пропусков, дубликатов, поменяли тип поля messages и calls на целое число (int), т.к. дробных значений сообщений и звонков не бывает.
- Для исследования потребовалось разбить выборку на обучающую, валидационную и тестовую.
- Были исследованы модели Decision Tree, RandomForest и Logistic Regression;
- Наилучшую эффективность (точность/скорость) показала модель RandomForest с гиперпараметрами n_estimators равным - 8 и max_depth равным 8;
- модель RandomForest с наилучшими гиперапараметрами была проверена на тестовой выборке и показала результат Accuracy - 0.7962674961119751;
- модель RandomForest проверенная на тестовой выборке была проверена на вменяемость - для проверки использовалась модель DummyClassifier, которая показала результат Accuracy - 0.7060653188180405.